### PR TITLE
1510 api for creating program requisitions 

### LINKED ISSUE #1510: Program Requisitions - API for creating program requisition
From epic: #1451 

Implement this API on back end:

```graphql
InsertProgramRequestRequisition(input: insertProgramRequestRequisitionInput) {
// result similar to normal import
}

insertProgramRequestRequisitionInput {
   otherPartyId: String
   programOrderTypeId: String
   periodId: String
}
```

### Errors

There is a chance that two orders are created in concurrently, using up limit of order for order type and period, so a Structured error should be returned for that, otherwise standard user input error should be fine for other errors.

### Other

And program settings can be deleted and changed by central server, so we would only store program_order_type_name against program.

maxMOS and thresholdMOS should carry over from  programOrderType

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #1510 

# 👩🏻‍💻 What does this PR do? 
New graphql request `insertProgramRequestRequisition`
this takes in additional parameters `programOrderTypeId` and `periodId`

`program_id`, `max_mos`, `min_mos` and `order_type` are set on the requisition based on the new fields.

Follow up after merging other PRs #1599

# 🧪 How has/should this change been tested? 

Will be best tested when integrated with #1509

## 💌 Any notes for the reviewer?

I've copied most of the generation and validation logic from `insert` for request requisition.
Maybe some of that should be extracted so code can be shared between the two? Not entirely clear what kinds of things will be different between the types at this stage.

Should there be any additional validations added for program requistions

## 📃 Documentation

_No user facing changes_

### CODE REVIEW

**Review by mark-prins — APPROVED** [review]
👍  Looks good to me. 
Have run tests and added a requisition and tested invalid data / error handling. working fine for me

File: server/service/src/requisition/request_requisition/insert_program.rs [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+fn validate(
+    connection: &StorageConnection,
+    store_id: &str,
+    input: &InsertProgramRequestRequisition,
+) -> Result<(), OutError> {
+    if let Some(_) = check_requisition_exists(connection, &input.id)? {
+        return Err(OutError::RequisitionAlreadyExists);
+    }
+
+    let other_party = check_other_party(
```
Comment (andreievg) [inline server/service/src/requisition/request_requisition/insert_program.rs:68]: Sorry, I should have explained a little bit further, this should technically be verified with:

https://github.com/openmsupply/open-msupply/blob/692b37d560cb2f495f3b2d38c86e2310f11b03f2/server/service/src/requisition/program_settings/mod.rs#L31

against: https://github.com/openmsupply/open-msupply/blob/692b37d560cb2f495f3b2d38c86e2310f11b03f2/server/service/src/requisition/program_settings/mod.rs#L23

Comment (andreievg) [reply server/service/src/requisition/request_requisition/insert_program.rs:68]: Also order_type and program_id would be checked against that struct, i see you have a todo for that, thanks !

File: server/service/src/requisition/request_requisition/insert_program.rs [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+    // Lookup order type
+    let order_type = ProgramRequisitionOrderTypeRowRepository::new(connection)
+        .find_one_by_id(&program_order_type_id)?;
+    let order_type = match order_type {
+        Some(order_type) => order_type,
+        None => return Err(RepositoryError::NotFound),
+    };
+
+    // Lookup program requisition settings
+    let program_requisition_settings = ProgramRequisitionSettingsRowRepository::new(connection)
```
Comment (andreievg) [inline server/service/src/requisition/request_requisition/insert_program.rs:114]: As per the other comment, we could use ProgramSettings here (service result)

**Review by andreievg — APPROVED** [review]
This looks great, thanks @jmbrunskill, I think it's ok to merge this in and then do the extra changes when https://github.com/openmsupply/open-msupply/pull/1595 is merged, or you can branch from https://github.com/openmsupply/open-msupply/pull/1595 and update from develop (if you want to get going on it sooner).

**Pull request merged**